### ISSUE TITLE
"half typed" dict literal is slow since pyright@1.1.229

### ISSUE THREAD

**Issue by arjennienhuis** [microsoft/pyright#3274]
**Describe the bug**
My config dict that I typed as `dict[str, list]` is making pyright very slow since 1.1.229.

**To Reproduce**
```shell
$ python3 -c "print('CONFIG: dict[str, list] = ', {str(a): [(str(a), a)[:n % 3] for n in range(a % 7)] for a in range(100)})" > slow.py
$ pyright --version
pyright 1.1.234
$ pyright --stats slow.py
No configuration file found.
No pyproject.toml file found.
Assuming Python platform Linux
Searching for source files
Found 1 source file
0 errors, 0 warnings, 0 informations 
Completed in 32.99sec

Analysis stats
Total files analyzed: 19

Timing stats
Find Source Files:    0sec
Read Source Files:    0.01sec
Tokenize:             0.09sec
Parse:                0.11sec
Resolve Imports:      0.05sec
Bind:                 0.07sec
Check:                32.32sec
Detect Cycles:        0sec
$ for v in pyright@1.1.{217..234}; do
> npm install -g "$v" > /dev/null
> echo -ne "$v | "; /bin/time  -f %e pyright slow2.py > /dev/null
> done
pyright@1.1.217 | 1.57
pyright@1.1.218 | 1.49
pyright@1.1.219 | 3.25
pyright@1.1.220 | 3.26
pyright@1.1.221 | 3.51
pyright@1.1.222 | 3.37
pyright@1.1.223 | 3.38
pyright@1.1.224 | 3.52
pyright@1.1.225 | 3.54
pyright@1.1.226 | 3.39
pyright@1.1.227 | 3.53
pyright@1.1.228 | 3.48
pyright@1.1.229 | 30.75
pyright@1.1.230 | 30.77
pyright@1.1.231 | 31.42
pyright@1.1.232 | 30.81
pyright@1.1.233 | 31.65
pyright@1.1.234 | 33.24

```

**Expected behavior**
I expected pyright to be faster in this specific case

**Screenshots or Code**
```
$ head fast1.py fast2.py slow.py 
==> fast1.py <==
CONFIG: dict[str, list[tuple[()] | tuple[str] | tuple[str, int]]] = {
    "0": [],
    "1": [()],
    "2": [(), ("2",)],
    "3": [(), ("3",), ("3", 3)],
    "4": [(), ("4",), ("4", 4), ()],
    "5": [(), ("5",), ("5", 5), (), ("5",)],
    "6": [(), ("6",), ("6", 6), (), ("6",), ("6", 6)],
    "7": [],
    "8": [()],

==> fast2.py <==
CONFIG = {
    "0": [],
    "1": [()],
    "2": [(), ("2",)],
    "3": [(), ("3",), ("3", 3)],
    "4": [(), ("4",), ("4", 4), ()],
    "5": [(), ("5",), ("5", 5), (), ("5",)],
    "6": [(), ("6",), ("6", 6), (), ("6",), ("6", 6)],
    "7": [],
    "8": [()],

==> slow.py <==
CONFIG: dict[str, list] = {
    "0": [],
    "1": [()],
    "2": [(), ("2",)],
    "3": [(), ("3",), ("3", 3)],
    "4": [(), ("4",), ("4", 4), ()],
    "5": [(), ("5",), ("5", 5), (), ("5",)],
    "6": [(), ("6",), ("6", 6), (), ("6",), ("6", 6)],
    "7": [],
    "8": [()],

$ pyright --verbose --stats fast1.py fast2.py slow.py 
No configuration file found.
No pyproject.toml file found.
Assuming Python platform Linux
Search paths for /home/arjen/tmp/test
  /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib
  /home/arjen/tmp/test
  /home/arjen/tmp/test/typings
  /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/lib/python3.9
  /usr/lib/python3.9/lib-dynload
  /home/arjen/.local/lib/python3.9/site-packages
  /usr/local/lib/python3.9/dist-packages
  /usr/lib/python3/dist-packages
  /usr/lib/python3.9/dist-packages
Searching for source files
Found 3 source files
[FG] analyzing: /home/arjen/tmp/test/fast1.py ...
[FG]   parsing: /home/arjen/tmp/test/fast1.py [fs read 2ms] (59ms)
[FG]   parsing: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 3ms] (115ms)
[FG]   binding: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/builtins.pyi (49ms)
[FG]   binding: /home/arjen/tmp/test/fast1.py (2ms)
[FG]   checking: /home/arjen/tmp/test/fast1.py ...
[FG]     evaluateTypesForStatement ...
[FG]       import lookup ...
[FG]         parsing: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/typing_extensions.pyi [fs read 2ms] (14ms)
[FG]         binding: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/typing_extensions.pyi (2ms)
[FG]       import lookup [typing_extensions] (16ms) [f:2, t:4, p:2, i:2, b:2]
[FG]       import lookup ...
[FG]         parsing: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/typing.pyi [fs read 0ms] (40ms)
[FG]         binding: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/typing.pyi (12ms)
[FG]       import lookup [typing] (53ms) [f:0, t:14, p:24, i:2, b:12]
[FG]       import lookup ...
[FG]         parsing: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi [fs read 0ms] (11ms)
[FG]         binding: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi (7ms)
[FG]       import lookup [_typeshed] (18ms) [f:0, t:1, p:2, i:8, b:7]
[FG]       import lookup ...
[FG]         parsing: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/types.pyi [fs read 0ms] (26ms)
[FG]         binding: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/types.pyi (7ms)
[FG]       import lookup [types] (34ms) [f:0, t:12, p:10, i:4, b:7]
[FG]       import lookup ...
[FG]         parsing: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/abc.pyi [fs read 0ms] (2ms)
[FG]         binding: /home/arjen/.npm-packages/lib/node_modules/pyright/dist/typeshed-fallback/stdlib/abc.pyi (3ms)
[FG]       import lookup [abc] (5ms) [f:0, t:0, p:1, i:0, b:3]
[FG]     evaluateTypesForStatement [Assignment (fast1)] (255ms) [f:2, t:31, p:39, i:16, b:31]
[FG]   checking: /home/arjen/tmp/test/fast1.py (267ms)
[FG] analyzing: /home/arjen/tmp/test/fast1.py (495ms)
[FG] analyzing: /home/arjen/tmp/test/fast2.py ...
[FG]   parsing: /home/arjen/tmp/test/fast2.py [fs read 0ms] (4ms)
[FG]   binding: /home/arjen/tmp/test/fast2.py (1ms)
[FG]   checking: /home/arjen/tmp/test/fast2.py (16ms)
[FG] analyzing: /home/arjen/tmp/test/fast2.py (22ms)
[FG] analyzing: /home/arjen/tmp/test/slow.py ...
[FG]   parsing: /home/arjen/tmp/test/slow.py [fs read 0ms] (12ms)
[FG]   binding: /home/arjen/tmp/test/slow.py (3ms)
[FG]   checking: /home/arjen/tmp/test/slow.py ...
[FG]     evaluateTypesForStatement [Assignment (slow)] (31254ms)
[FG] Long operation: evaluateTypesForStatement (31254ms)
[FG]   checking: /home/arjen/tmp/test/slow.py (31258ms)
[FG] Long operation: checking: /home/arjen/tmp/test/slow.py (31258ms)
[FG] analyzing: /home/arjen/tmp/test/slow.py (31274ms)
[FG] Long operation: analyzing: /home/arjen/tmp/test/slow.py (31274ms)
0 errors, 0 warnings, 0 informations 
Completed in 32.115sec

Analysis stats
Total files analyzed: 21

Timing stats
Find Source Files:    0sec
Read Source Files:    0.01sec
Tokenize:             0.1sec
Parse:                0.12sec
Resolve Imports:      0.04sec
Bind:                 0.09sec
Check:                31.42sec
Detect Cycles:        0sec


```

**Comment by erictraut** [comment]
Thanks for the analysis. Could you provide the full "slow.py" file or describe its contents so I can accurately repro the problem? I presume that the dict literal is very long? I was able to repro long analysis times, but only after creating a dict literal with tens of thousands of entries.

**Comment by arjennienhuis** [comment]
It's not tens of thousands:

```python
CONFIG: dict[str, list] = {
    "0": [],
    "1": [()],
    "2": [(), ("2",)],
    "3": [(), ("3",), ("3", 3)],
    "4": [(), ("4",), ("4", 4), ()],
    "5": [(), ("5",), ("5", 5), (), ("5",)],
    "6": [(), ("6",), ("6", 6), (), ("6",), ("6", 6)],
    "7": [],
    "8": [()],
    "9": [(), ("9",)],
    "10": [(), ("10",), ("10", 10)],
    "11": [(), ("11",), ("11", 11), ()],
    "12": [(), ("12",), ("12", 12), (), ("12",)],
    "13": [(), ("13",), ("13", 13), (), ("13",), ("13", 13)],
    "14": [],
    "15": [()],
    "16": [(), ("16",)],
    "17": [(), ("17",), ("17", 17)],
    "18": [(), ("18",), ("18", 18), ()],
    "19": [(), ("19",), ("19", 19), (), ("19",)],
    "20": [(), ("20",), ("20", 20), (), ("20",), ("20", 20)],
    "21": [],
    "22": [()],
    "23": [(), ("23",)],
    "24": [(), ("24",), ("24", 24)],
    "25": [(), ("25",), ("25", 25), ()],
    "26": [(), ("26",), ("26", 26), (), ("26",)],
    "27": [(), ("27",), ("27", 27), (), ("27",), ("27", 27)],
    "28": [],
    "29": [()],
    "30": [(), ("30",)],
    "31": [(), ("31",), ("31", 31)],
    "32": [(), ("32",), ("32", 32), ()],
    "33": [(), ("33",), ("33", 33), (), ("33",)],
    "34": [(), ("34",), ("34", 34), (), ("34",), ("34", 34)],
    "35": [],
    "36": [()],
    "37": [(), ("37",)],
    "38": [(), ("38",), ("38", 38)],
    "39": [(), ("39",), ("39", 39), ()],
    "40": [(), ("40",), ("40", 40), (), ("40",)],
    "41": [(), ("41",), ("41", 41), (), ("41",), ("41", 41)],
    "42": [],
    "43": [()],
    "44": [(), ("44",)],
    "45": [(), ("45",), ("45", 45)],
    "46": [(), ("46",), ("46", 46), ()],
    "47": [(), ("47",), ("47", 47), (), ("47",)],
    "48": [(), ("48",), ("48", 48), (), ("48",), ("48", 48)],
    "49": [],
    "50": [()],
    "51": [(), ("51",)],
    "52": [(), ("52",), ("52", 52)],
    "53": [(), ("53",), ("53", 53), ()],
    "54": [(), ("54",), ("54", 54), (), ("54",)],
    "55": [(), ("55",), ("55", 55), (), ("55",), ("55", 55)],
    "56": [],
    "57": [()],
    "58": [(), ("58",)],
    "59": [(), ("59",), ("59", 59)],
    "60": [(), ("60",), ("60", 60), ()],
    "61": [(), ("61",), ("61", 61), (), ("61",)],
    "62": [(), ("62",), ("62", 62), (), ("62",), ("62", 62)],
    "63": [],
    "64": [()],
    "65": [(), ("65",)],
    "66": [(), ("66",), ("66", 66)],
    "67": [(), ("67",), ("67", 67), ()],
    "68": [(), ("68",), ("68", 68), (), ("68",)],
    "69": [(), ("69",), ("69", 69), (), ("69",), ("69", 69)],
    "70": [],
    "71": [()],
    "72": [(), ("72",)],
    "73": [(), ("73",), ("73", 73)],
    "74": [(), ("74",), ("74", 74), ()],
    "75": [(), ("75",), ("75", 75), (), ("75",)],
    "76": [(), ("76",), ("76", 76), (), ("76",), ("76", 76)],
    "77": [],
    "78": [()],
    "79": [(), ("79",)],
    "80": [(), ("80",), ("80", 80)],
    "81": [(), ("81",), ("81", 81), ()],
    "82": [(), ("82",), ("82", 82), (), ("82",)],
    "83": [(), ("83",), ("83", 83), (), ("83",), ("83", 83)],
    "84": [],
    "85": [()],
    "86": [(), ("86",)],
    "87": [(), ("87",), ("87", 87)],
    "88": [(), ("88",), ("88", 88), ()],
    "89": [(), ("89",), ("89", 89), (), ("89",)],
    "90": [(), ("90",), ("90", 90), (), ("90",), ("90", 90)],
    "91": [],
    "92": [()],
    "93": [(), ("93",)],
    "94": [(), ("94",), ("94", 94)],
    "95": [(), ("95",), ("95", 95), ()],
    "96": [(), ("96",), ("96", 96), (), ("96",)],
    "97": [(), ("97",), ("97", 97), (), ("97",), ("97", 97)],
    "98": [],
    "99": [()],
}
```

**Comment by erictraut** [comment]
Thanks, this will be fixed in the next release. One of the bug fixes I introduced in 1.1.229 added an n^2 behavior in certain cases. You were hitting this case with a relatively large value of n. I've fixed this so it's linear in most cases.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.235, which I just pblished. It will also be included in the next release of pylance.

**Comment by arjennienhuis** [comment]
It's fixed. Thanks! With my real data I get:

| version         | time (s) |
| :-------------- | -------: |
| pyright@1.1.218 |    30.27 |
| ...             |          |
| pyright@1.1.228 |    35.52 |
| pyright@1.1.229 |   133.00 |
| ...             |          |
| pyright@1.1.234 |   121.34 |
| pyright@1.1.235 |     3.18 |
